### DISCUSSION TITLE
Has `ResponseBuilder::json<T>` been broken?

### DISCUSSION THREAD

**Discussion by Overdash** [actix/actix-web#2058] [Q&A] [answered]
I updated from actix-3.3 to 4.0.0-beta.3 (because I desperately need the Tokio v1) and my code (expectedly) has been broken.
However, it seems I can no longer pass arbitrary Structs which derive `serde::Serialize` into the `ResponseBuilder::json<T>()`?

for example if I do
```
HttpResponse::Ok().json(VerificationCheckResponse{
        verified: user.email_verified
    })
```

Where `VerificationCheckResponse` is 
```rust
#[derive(Serialize)]
pub struct VerificationCheckResponse {
    verified: bool
}
```

I get `error[E0277]: the trait bound `VerificationCheckResponse: Deref` is not satisfied` :
```rust
  HttpResponse::Ok().json(VerificationCheckResponse{
   |  ____________________________________^
56 | |         verified: user.email_verified
57 | |     })
   | |_____^ the trait `Deref` is not implemented for `VerificationCheckResponse`
```

I checked the actix-http code and found `json<T>` signature has been changed from simply 
```rust
pub fn json<T: Serialize>(&mut self, value: T) -> Response {...}
```
to 

```rust
pub fn json<T>(&mut self, value: T) -> Response
    where
        T: ops::Deref,
        T::Target: Serialize,
{...}
```
 It doesn't seem right for us to implement Deref for every struct we wish to return as a basic JSON, also it seriously hurts developer QoL and overall ergonomics of the framework if it is intentional (imo, ofc).

So, is this intentional?
Perhaps I'm missing something; any insight?

**Comment by robjtede** [comment, marked as answer]
The easy fix is to just pass `&VerificationCheckResponse {...}`.

While the change to take `T: ops::Deref` was intentional, it was deemed slightly better to use a simpler approach and impl Serialize in more of our own structs for better DX. See https://github.com/actix/actix-web/pull/2052.

**Reply by Overdash** [reply]
Wow - thanks for the quick reply. 
I understand better now makes sense. After brainfarting all over my keyboard I forgot I could just pass it AsRef. My bad lol.
Thanks again :)
